In [82]:
import pandas as pd
import random 
data = pd.read_csv('https://s3-eu-west-1.amazonaws.com/shanebucket/downloads/uk-500.csv')
# assigning id for employees randomly
data['id'] = [random.randint(0,1000) for x in range(data.shape[0])]
data

first_name   last_name                    company_name  \
0        Aleshia  Tomkiewicz         Alan D Rosenburg Cpa Pc   
1           Evan   Zigomalas              Cap Gemini America   
2         France     Andrade             Elliott, John W Esq   
3        Ulysses   Mcwalters                  Mcmahan, Ben L   
4         Tyisha      Veness                  Champagne Room   
5           Eric       Rampy         Thompson, Michael C Esq   
6           Marg    Grasmick   Wrangle Hill Auto Auct & Slvg   
7        Laquita       Hisaw           In Communications Inc   
8           Lura    Manzella                 Bizerba Usa Inc   
9         Yuette      Klapec                       Max Video   
10      Fernanda      Writer            K & R Associates Inc   
11   Charlesetta         Erm                Cain, John M Esq   
12      Corrinne       Jaret               Sound Vision Corp   
13        Niesha       Bruch          Rowley/hansell Petetin   
14        Rueben   Gastellum   Industrial Engineering Assocs   
15       Michell   Throssell             Weiss Spirt & Guyer   
16         Edgar       Kanne           Crowan, Kenneth W Esq   
17        Dewitt       Julio            Rittenhouse Motor Co   
18      Charisse    Spinello            Modern Plastics Corp   
19           Mee    Lapinski        Galloway Electric Co Inc   
20         Peter   Gutierres      Niagara Custombuilt Mfg Co   
21       Octavio   Salvadore            Practical Periphrals   
22        Martha     Teplica          Curtin, Patricia M Esq   
23       Tamesha      Veigel             Wilhelm, James E Jr   
24          Tess       Sitra                     Smart Signs   
25       Leonard      Kufner    Arctic Star Distributing Inc   
26      Svetlana      Tauras        Lensbower, Gregory L Esq   
27           Pok    Molaison           Portal Metal Products   
28     Augustine    Growcock               Lynema, Cliff Cpa   
29         Karma      Quarto                 J C S Machinery   
..           ...         ...                             ...   
470         Tony  Diazdeleon              Dilling Harris Inc   
471          Val      Villot            Horizon Air Magazine   
472          Mel    Picciuto                  Avante Limited   
473     Isabella  Piatkowski           Miller, William E Iii   
474       Erasmo        Rhea                Martin Morrissey   
475        Ivory     Lohrenz                 Auto Body Color   
476       Nikita       Walka               Fred, Anthony Esq   
477        Aleta      Ligons           Advantage Bookkeeping   
478         Owen     Jentzen           Capital Fasteners Inc   
479      Pauline       Fling      Cumberland Vacuum Products   
480      Tijuana    Machalek          Schwartz, Thomas E Esq   
481        Ahmad     Alsaqri    Alliance Construction Co Inc   
482       Jamika      Conoly               Keller, Raymond C   
483      Derrick     Dolloff         Proto Tek Manufacturing   
484   Jacquelyne     Reibman            Great Clips For Hair   
485      Zachary  Freeburger                 Country Kitchen   
486       Sophia     Gaucher               T C E Systems Inc   
487     Isabelle        Kono    Rock Springs Petroleum Equip   
488       Ronnie     Brigman                  Howard Johnson   
489     Krystina  Schlabaugh           Harris Oldsmobile Inc   
490       Rosita   Ausdemore               Jurdem, Scott Esq   
491         Huey     Stancil                   Lindner Funds   
492       Elbert     Fiorino            Donald, G Nelson Esq   
493         Pura     Manciel                  Cia Inspection   
494       Alesia       Katie    Nelsons Trmt & Pest Cntrl Co   
495        Avery        Veit      Plaza Gourmet Delicatessen   
496         Reid     Euresti            Fitzgerald, Edward J   
497    Charlette    Brenning              Furey & Associates   
498    Celestina       Keeny        Bfg Federal Credit Union   
499           Mi      Richan  Nelson Wright Haworth Golf Crs   

                    address          

In [14]:
data.shape

(500, 12)

# 1.Selecting pandas data using iloc 

In [30]:
# selecting columns
data.iloc[:,1]
data.iloc[0:2]
# selecting rows
data.iloc[0]
data.iloc[:,0:2]

first_name                                   Aleshia
last_name                                 Tomkiewicz
company_name                 Alan D Rosenburg Cpa Pc
address                                 14 Taylor St
city                               St. Stephens Ward
county                                          Kent
postal                                       CT2 7PP
phone1                                  01835-703597
phone2                                  01944-369967
email                        atomkiewicz@hotmail.com
web             http://www.alandrosenburgcpapc.co.uk
id                                               299
Name: 0, dtype: object

In [37]:
data.iloc[[0,1,2,3]]
data.iloc[[1,5,2,3,6],[5,2,6,3]]

county                   company_name    postal           address
1   Buckinghamshire             Cap Gemini America  HP11 2AX       5 Binney St
5  Northamptonshire        Thompson, Michael C Esq  NN14 2GH      9472 Lind St
2       Bournemouth            Elliott, John W Esq   BH6 3BE      8 Moor Place
3      Lincolnshire                 Mcmahan, Ben L  DN36 5RP     505 Exeter Rd
6       Southampton  Wrangle Hill Auto Auct & Slvg  SO14 3TY  7457 Cowl St #70

# 2. Selecting pandas using 'loc'

In [79]:
# 2(a) : Label based
# setting first_name as label and selecting data
data.set_index('first_name', inplace=True)
data.head(5)

last_name             company_name            address  \
first_name                                                           
Aleshia     Tomkiewicz  Alan D Rosenburg Cpa Pc       14 Taylor St   
Evan         Zigomalas       Cap Gemini America        5 Binney St   
France         Andrade      Elliott, John W Esq       8 Moor Place   
Ulysses      Mcwalters           Mcmahan, Ben L      505 Exeter Rd   
Tyisha          Veness           Champagne Room  5396 Forth Street   

                                      city           county    postal  \
first_name                                                              
Aleshia                  St. Stephens Ward             Kent   CT2 7PP   
Evan                            Abbey Ward  Buckinghamshire  HP11 2AX   
France      East Southbourne and Tuckton W      Bournemouth   BH6 3BE   
Ulysses                 Hawerby cum Beesby     Lincolnshire  DN36 5RP   
Tyisha          Greets Green and Lyng Ward    West Midlands   B70 9DT   

                  phone1        phone2                       email  \
first_name                                                           
Aleshia     01835-703597  01944-369967     atomkiewicz@hotmail.com   
Evan        01937-864715  01714-737668    evan.zigomalas@gmail.com   
France      01347-368222  01935-821636  france.andrade@hotmail.com   
Ulysses     01912-771311  01302-601380         ulysses@hotmail.com   
Tyisha      01547-429341  01290-367248   tyisha.veness@hotmail.com   

                                             web   id  
first_name                                             
Aleshia     http://www.alandrosenburgcpapc.co.uk  783  
Evan           http://www.capgeminiamerica.co.uk   20  
France          http://www.elliottjohnwesq.co.uk  821  
Ulysses             http://www.mcmahanbenl.co.uk  422  
Tyisha            http://www.champagneroom.co.uk  721

for selecting using row value, that row needs to be the index of that dataset

In [80]:
# select specific record from dataset using emp name
data.loc['Evan'] # single employee
data.loc[['Evan','France']] # multiple employees
data.loc[['Evan','France'],['city','county']] # select multiple employees with specific columns
data.loc[['Evan','France','Erasmo'],'city':'phone2']

city           county    postal  \
first_name                                                              
Evan                            Abbey Ward  Buckinghamshire  HP11 2AX   
France      East Southbourne and Tuckton W      Bournemouth   BH6 3BE   
Erasmo                    Darton West Ward  South Yorkshire   S75 5JY   
Erasmo                         Sunningdale        Berkshire   SL5 0AZ   
Erasmo                            Mortehoe            Devon  EX34 7DQ   

                  phone1        phone2  
first_name                              
Evan        01937-864715  01714-737668  
France      01347-368222  01935-821636  
Erasmo      01492-454455  01304-185373  
Erasmo      01445-796544  01650-307192  
Erasmo      01507-386397  01899-316385

Note that in the last example, data.loc[487] (the row with index value 487) is not equal to data.iloc[487] (the 487th row in the data). The index of the DataFrame can be out of numeric order, and/or a string or multi-value.

In [89]:
# 2 (b) : Boolean/logical indexing using .loc
data.loc[data['first_name'] == 'Erasmo','city':'email'] # select's columns between city and email 
data.loc[data['first_name'] == 'Erasmo',['city','email']] # select's only city, email columns

city                         email
233  Darton West Ward  erasmo.talentino@hotmail.com
393       Sunningdale             egath@hotmail.com
474          Mortehoe       erasmo_rhea@hotmail.com

# 3. Select Pandas data using ix 

1. it is a hybrid of .loc and .iloc
2. .loc can be replaced .ix when string is passed to it.
        eg:  data.ix[['Andrade']] == data.loc[['Andrade']]
3. loc can be replaced with .ix when int is passed to it.
        eg: data.ix[[33]] == data.iloc[[33]]

In [101]:
# change values from dataFrame using .loc
# changes first_name to Evan if id > 0
data.loc[data['id'] > 0, 'first_name'] = "Evan"
data

first_name   last_name                    company_name  \
0         Evan  Tomkiewicz         Alan D Rosenburg Cpa Pc   
1         Evan   Zigomalas              Cap Gemini America   
2         Evan     Andrade             Elliott, John W Esq   
3         Evan   Mcwalters                  Mcmahan, Ben L   
4         Evan      Veness                  Champagne Room   
5         Evan       Rampy         Thompson, Michael C Esq   
6         Evan    Grasmick   Wrangle Hill Auto Auct & Slvg   
7         Evan       Hisaw           In Communications Inc   
8         Evan    Manzella                 Bizerba Usa Inc   
9         Evan      Klapec                       Max Video   
10        Evan      Writer            K & R Associates Inc   
11        Evan         Erm                Cain, John M Esq   
12        Evan       Jaret               Sound Vision Corp   
13        Evan       Bruch          Rowley/hansell Petetin   
14        Evan   Gastellum   Industrial Engineering Assocs   
15        Evan   Throssell             Weiss Spirt & Guyer   
16        Evan       Kanne           Crowan, Kenneth W Esq   
17        Evan       Julio            Rittenhouse Motor Co   
18        Evan    Spinello            Modern Plastics Corp   
19        Evan    Lapinski        Galloway Electric Co Inc   
20        Evan   Gutierres      Niagara Custombuilt Mfg Co   
21        Evan   Salvadore            Practical Periphrals   
22        Evan     Teplica          Curtin, Patricia M Esq   
23        Evan      Veigel             Wilhelm, James E Jr   
24        Evan       Sitra                     Smart Signs   
25        Evan      Kufner    Arctic Star Distributing Inc   
26        Evan      Tauras        Lensbower, Gregory L Esq   
27        Evan    Molaison           Portal Metal Products   
28        Evan    Growcock               Lynema, Cliff Cpa   
29        Evan      Quarto                 J C S Machinery   
..         ...         ...                             ...   
470       Evan  Diazdeleon              Dilling Harris Inc   
471       Evan      Villot            Horizon Air Magazine   
472       Evan    Picciuto                  Avante Limited   
473       Evan  Piatkowski           Miller, William E Iii   
474       Evan        Rhea                Martin Morrissey   
475       Evan     Lohrenz                 Auto Body Color   
476       Evan       Walka               Fred, Anthony Esq   
477       Evan      Ligons           Advantage Bookkeeping   
478       Evan     Jentzen           Capital Fasteners Inc   
479       Evan       Fling      Cumberland Vacuum Products   
480       Evan    Machalek          Schwartz, Thomas E Esq   
481       Evan     Alsaqri    Alliance Construction Co Inc   
482       Evan      Conoly               Keller, Raymond C   
483       Evan     Dolloff         Proto Tek Manufacturing   
484       Evan     Reibman            Great Clips For Hair   
485       Evan  Freeburger                 Country Kitchen   
486       Evan     Gaucher               T C E Systems Inc   
487       Evan        Kono    Rock Springs Petroleum Equip   
488       Evan     Brigman                  Howard Johnson   
489       Evan  Schlabaugh           Harris Oldsmobile Inc   
490       Evan   Ausdemore               Jurdem, Scott Esq   
491       Evan     Stancil                   Lindner Funds   
492       Evan     Fiorino            Donald, G Nelson Esq   
493       Evan     Manciel                  Cia Inspection   
494       Evan       Katie    Nelsons Trmt & Pest Cntrl Co   
495       Evan        Veit      Plaza Gourmet Delicatessen   
496       Evan     Euresti            Fitzgerald, Edward J   
497       Evan    Brenning              Furey & Associates   
498       Evan       Keeny        Bfg Federal Credit Union   
499       Evan      Richan  Nelson Wright Haworth Golf Crs   

                    address                            city  \
0              14 Taylor St               St. Stephens Ward   
1               5 Binney St      